Wrappers/Python/ccpi/optimisation/functions/MixedL21Norm.py

```diff
@@ -133,35 +92,57 @@ def proximal(self, x, tau, out=None):
         
         if out is None:
             
+        
             tmp = x.pnorm(2)
             res = (tmp - tau).maximum(0.0) * x/tmp
             
+            # TODO avoid using numpy, add operation in the framework
+            # This will be useful when we add cupy 
+                                 
             for el in res.containers:
-                el.as_array()[np.isnan(el.as_array())]=0            
+                
+                elarray = el.as_array()
+                elarray[np.isnan(elarray)]=0
+                el.fill(elarray)                
+                
+#                el.as_array()[np.isnan(el.as_array())]=0            
             
             return res            
             
         else:
             
-            tmp = x.pnorm(2)
-            res = (tmp - tau).maximum(0.0) * x/tmp
-
-            for el in res.containers:
-                el.as_array()[np.isnan(el.as_array())]=0
-
-            out.fill(res)            
-            
-                        
-#            tmp = functools.reduce(lambda a,b: a + b*b, x.containers, x.get_item(0) * 0 ).sqrt()
+#            tmp = x.pnorm(2)
 #            res = (tmp - tau).maximum(0.0) * x/tmp
 #
 #            for el in res.containers:
-#                el.as_array()[np.isnan(el.as_array())]=0
+#                
+#                elarray = el.as_array()
+#                elarray[np.isnan(elarray)]=0
+#                el.fill(elarray)  
 #
-#            out.fill(res)
+#            out.fill(res)   
+            
+            
+            tmp = x.pnorm(2)
+            tmp_ig = 0.0 * tmp
+            (tmp - tau).maximum(0.0, out = tmp_ig)
+            tmp_ig.multiply(x, out = out)
+            out.divide(tmp, out = out)
+            
+            for el in out.containers:
+                
+                elarray = el.as_array()
+                elarray[np.isnan(elarray)]=0
+                el.fill(elarray)  
+
+            out.fill(out)              
+            
+            
         
+# TODO, add the prox conjugate in the documenataion and then delete the code below
+            
+##############################################################################
 ##############################################################################
-        ##############################################################################
 #    def proximal_conjugate(self, x, tau, out=None): 
```

Comment:
In general please tidy up the file, removing commented old code and making use of blank lines consistent.